octopusdeploy_framework/datasource_tenant_projects.go

```diff
@@ -0,0 +1,201 @@
+package octopusdeploy_framework
+
+import (
+	"context"
+	"fmt"
+	"github.com/OctopusDeploy/go-octopusdeploy/v2/pkg/client"
+	"github.com/OctopusDeploy/go-octopusdeploy/v2/pkg/tenants"
+	"github.com/OctopusDeploy/terraform-provider-octopusdeploy/octopusdeploy_framework/schemas"
+	"github.com/OctopusDeploy/terraform-provider-octopusdeploy/octopusdeploy_framework/util"
+	"github.com/hashicorp/terraform-plugin-framework/attr"
+	"github.com/hashicorp/terraform-plugin-framework/datasource"
+	datasourceSchema "github.com/hashicorp/terraform-plugin-framework/datasource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"golang.org/x/exp/slices"
+)
+
+type TenantProjectsDataModel struct {
+	SpaceID        types.String `tfsdk:"space_id"`
+	TenantIDs      types.List   `tfsdk:"tenant_ids"`
+	ProjectIDs     types.List   `tfsdk:"project_ids"`
+	EnvironmentIDs types.List   `tfsdk:"environment_ids"`
+	TenantProjects types.List   `tfsdk:"tenant_projects"`
+}
+
+type tenantProjectsDataSource struct {
+	*Config
+}
+
+func NewTenantProjectDataSource() datasource.DataSource {
+	return &tenantProjectsDataSource{}
+}
+
+func (*tenantProjectsDataSource) Metadata(_ context.Context, req datasource.MetadataRequest, resp *datasource.MetadataResponse) {
+	resp.TypeName = util.GetTypeName("tenant_projects")
+}
+
+func (t *tenantProjectsDataSource) Configure(_ context.Context, req datasource.ConfigureRequest, resp *datasource.ConfigureResponse) {
+	t.Config = DataSourceConfiguration(req, resp)
+}
+
+func (*tenantProjectsDataSource) Schema(_ context.Context, req datasource.SchemaRequest, resp *datasource.SchemaResponse) {
+	resp.Schema = datasourceSchema.Schema{
+		Description: "Provides information about existing tenants.",
+		Attributes: map[string]datasourceSchema.Attribute{
+			"tenant_ids":      schemas.GetQueryIDsDatasourceSchema(),
+			"project_ids":     schemas.GetQueryIDsDatasourceSchema(),
+			"environment_ids": schemas.GetQueryIDsDatasourceSchema(),
+			"space_id":        schemas.GetSpaceIdDatasourceSchema("tenant projects", false),
+		},
+		Blocks: map[string]datasourceSchema.Block{
+			"tenant_projects": datasourceSchema.ListNestedBlock{
+				Description: "A list of related tenants, projects and environments that match the filter(s).",
+				NestedObject: datasourceSchema.NestedBlockObject{
+					Attributes: map[string]datasourceSchema.Attribute{
+						"id": schemas.GetIdDatasourceSchema(true),
+						"tenant_id": datasourceSchema.StringAttribute{
+							Description: "The tenant ID associated with this tenant.",
+							Computed:    true,
+						},
+						"project_id": datasourceSchema.StringAttribute{
+							Description: "The project ID associated with this tenant.",
+							Computed:    true,
+						},
+						"environment_ids": datasourceSchema.ListAttribute{
+							Description: "The environment IDs associated with this tenant.",
+							ElementType: types.StringType,
+							Computed:    true,
+						},
+					},
+				},
+			},
+		},
+	}
+}
```

Comment:
Probably for a future work, this schema definition along with the resource one could be moved to its own file under `schemas`, this makes it more manageable and avoid drift.